hie-plugin-api/Haskell/Ide/Engine/PluginDescriptor.hs

```diff
@@ -93,72 +105,75 @@ data Service = Service
 
-type Plugins = Map.Map PluginId PluginDescriptor
+type Plugins = Map.Map PluginId UntaggedPluginDescriptor
+
+untagCommand :: NamedCommand t -> UntaggedCommand
+untagCommand (NamedCommand _ (Command desc func)) =
+  Command (desc {cmdContexts =
+                   recordToList' fromSing
+                                 (cmdContexts desc)
+                ,cmdAdditionalParams =
+                   recordToList' untagParamDesc
+                                 (cmdAdditionalParams desc)})
+          func
 
 -- | Ideally a Command is defined in such a way that its CommandDescriptor
 -- can be exposed via the native CLI for the tool being exposed as well.
 -- Perhaps use Options.Applicative for this in some way.
-data Command = forall a. (ValidResponse a) => Command
-  { cmdDesc :: !CommandDescriptor
+data Command desc = forall a. (ValidResponse a) => Command
+  { cmdDesc :: !desc
   , cmdFunc :: !(CommandFunc a)
   }
 
-instance Show Command where
+type TaggedCommand cxts tags
+  = Command (TaggedCommandDescriptor cxts tags)
+type UntaggedCommand = Command UntaggedCommandDescriptor
+
+instance Show desc => Show (Command desc) where
   show (Command desc _func) = "(Command " ++ show desc ++ ")"
 
+data NamedCommand (t :: CommandType) where
+        NamedCommand ::
+            KnownSymbol s =>
+            Proxy s ->
+            TaggedCommand cxts tags ->
+            NamedCommand ('CommandType s cxts tags)
+
+data CommandType = CommandType Symbol [AcceptedContext] [ParamDescType]
+
 -- | Build a command, ensuring the command response type name and the command
 -- function match
-buildCommand :: forall a. (ValidResponse a)
+buildCommand :: forall a s cxts tags. (ValidResponse a, KnownSymbol s)
   => CommandFunc a
-  -> CommandName
+  -> Proxy s
   -> T.Text
   -> [T.Text]
-  -> [AcceptedContext]
-  -> [ParamDescription]
-  -> Command
+  -> Rec SAcceptedContext cxts
+  -> Rec SParamDescription tags
+  -> NamedCommand ( 'CommandType s cxts tags )
 buildCommand fun n d exts ctxs parm =
-  Command
-  { cmdDesc = CommandDesc
-      { cmdName = n
-      , cmdUiDescription = d
-      , cmdFileExtensions = exts
-      , cmdContexts = ctxs
-      , cmdAdditionalParams = parm
-      , cmdReturnType = T.pack $ show $ typeOf (undefined::a)
-      }
-  , cmdFunc = fun
-  }
+  NamedCommand n $
+  Command {cmdDesc =
+            CommandDesc {cmdName = T.pack $ symbolVal n
+                        ,cmdUiDescription = d
+                        ,cmdFileExtensions = exts
+                        ,cmdContexts = ctxs
+                        ,cmdAdditionalParams = parm
+                        ,cmdReturnType =
+                           T.pack $ show $ typeOf (undefined :: a)}
+          ,cmdFunc = fun}
 
 -- ---------------------------------------------------------------------
 
--- | For a given 'AcceptedContext', define the parameters that are required in
--- the corresponding 'IdeRequest'
-contextMapping :: AcceptedContext -> [ParamDescription]
```

Comment:
Not yet, this PR only uses the type info for servant and then drops all the tags and uses the untagged versions. We should definitely try to use those types elsewhere too, but imho this PR is big enough already so I would prefer doing that in multiple steps.